roachtest: more cleanup in version-upgrade

tbg opened the pull request:
The struct wrapping a fn was no longer needed. This was left over from
supporting old versions, where the version could not be auto-detected.
Avoid redundant code, improve logging.

Add steps to abort each upgrade and finalize it only later.
This reliably hit the error

> pq: schema change cannot be initiated in this version until the
> version upgrade is finalized

from #44732 (comment).

I fixed this by marking all DDL feature tests as 20.1+ only.

Fixes #44732

Release note: None

tbg committed:
roachtest: simplify acceptance/version-upgrade

The struct wrapping a fn was no longer needed. This was left over from
supporting old versions, where the version could not be auto-detected.

Release note: None

tbg committed:
binfetcher: remove uninteresting log

Release note: None

tbg committed:
roachtest: more cleanup in version-upgrade

Avoid redundant code, improve logging

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47065)
<!-- Reviewable:end -->

tbg committed:
roachtest: rename upgrade -> autoupgrade

The "upgrade" roachtest is renamed to "autoupgrade" and moved into a new
file. The reason is that this test isn't particularly well-maintained or
extensible, and so we want to "hide" it and favor the mixed-version test
harness left in upgrade.go more prominently.

Release note: None

tbg committed:
roachtest: git mv {,version}upgrade.go

Let the file name match the test name.

Release note: None

tbg committed:
roachtest: version-upgrade: roll back upgrades

Add steps to abort each upgrade and finalize it only later.
This reliably hit the error

> pq: schema change cannot be initiated in this version until the
> version upgrade is finalized

from https://github.com/cockroachdb/cockroach/issues/44732#issuecomment-607333663.

I fixed this by marking all DDL feature tests as 20.1+ only.

Release note: None

tbg committed:
roachtest: rename a var

Tickled my OCD bone.

Release note: None

tbg committed:
roachtest: speed up acceptance/version-upgrade

This test was creating and closing SQL connections all the time, which
took around a second each time. This really added up. Cache these conns,
which speeds up this test quite a bit, requiring waits only for binary
upload and restarts.

Release note: None

spaskob reviewed (approved):
> Reviewed 1 of 2 files at r1, 2 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47065)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob and @tbg)

___
*[pkg/cmd/roachtest/autoupgrade.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4EzU6s37punZKq4xRo:-M4EzU6s37punZKq4xRp:bksizu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

thanks for splitting this out in a separate file - from a reader's point of view now it's easier to understand now.
Most of my comments are to aid readability for somebody vaguely familiar with how roachtests work.
___
*[pkg/cmd/roachtest/autoupgrade.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ewu_Z3wBYalIHSfHm:-M4Ewu_Z3wBYalIHSfHn:b-c7zmge) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L47)):*
> ```Go
> 
> 		// NB: remove startArgsDontEncrypt across this file once we're not running
> 		// roachtest against v2.1 any more (which would start a v2.0 cluster here).
> ```

For my understanding why do we test v2.1 - do we still have clients that we support running this version?
___
*[pkg/cmd/roachtest/autoupgrade.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ex9tN9KxPkz4UIdiJ:-M4Ex9tN9KxPkz4UIdiK:bkrd1vs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L70)):*
> ```Go
> 		// in the pool and throws an error at the next client using it (instead of
> 		// transparently reconnecting).
> 		db.SetMaxIdleConns(0)
> ```

Do you mind running locally to check if this is still true.
___
*[pkg/cmd/roachtest/autoupgrade.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ey3-_4RZLSAUMu6LV:-M4Ey3-_4RZLSAUMu6LW:bbyyhgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L104)):*
> ```Go
> 		}
> 
> 		oldVersion, err = clusterVersion()
> ```

nit: Please add a comment on why we over-write the function arg here.
___
*[pkg/cmd/roachtest/autoupgrade.go, line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4F0hy64epUQiWCg9Dd:-M4F0hy64epUQiWCg9De:b-2qlb50) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L134)):*
> ```Go
> 				t.Fatal(err)
> 			}
> 			c.Put(ctx, cockroach, "./cockroach", c.Node(i))
> ```

what is the version of the new binary?
___
*[pkg/cmd/roachtest/autoupgrade.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ez7sp5oR1fvelGrGa:-M4Ez7sp5oR1fvelGrGb:b-fk6wg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L169)):*
> ```Go
> 
> 		// Now decommission and stop n3.
> 		// The decommissioned nodes should not prevent auto upgrade.
> ```

nit: name the node since `nodes-2` is used quite a bit further in the code and it relies on a comment, ie

`decommissionedNode := nodes - 2`
___
*[pkg/cmd/roachtest/autoupgrade.go, line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4F-6AA731Fl-fo2Vgs:-M4F-6AA731Fl-fo2Vgt:b-rplhv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L170)):*
> ```Go
> 		// Now decommission and stop n3.
> 		// The decommissioned nodes should not prevent auto upgrade.
> 		if err := decommissionAndStop(nodes - 2); err != nil {
> ```

nit: add a comment explaining what do we test by decomissioning this node.
___
*[pkg/cmd/roachtest/versionupgrade.go, line 218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4F3BKhAhkPd5X9B1l5:-M4F3BKhAhkPd5X9B1l6:b8nek4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/versionupgrade.go#L218)):*
> ```Go
> }
> 
> func (u *versionUpgradeTest) uploadVersion(ctx context.Context, t *test, newVersion string) option {
> ```

Do you think these may be re-used by other mixed-version roachtests?


<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: skip autoupgrade in 2.1

2.1 is old and by skipping it we get to run this test with encryption at
rest.

Release note: None

tbg committed:
roachtest: rermove unnecessary hack

Release note: None

tbg committed:
roachtest: give autoupgrade a round of comments

It's in more dire need of reworking, but also it's unlikely that anyone
will depend on it, so explaining it a little better give the best ROI
for now.

Release note: None

tbg committed:
roachtest: fix name for autoupgrade

It's supposed to be called `autoupgrade` now.

Release note: None

tbg reviewed (commented):
TFTR @spaskob! Lmk if you have more comments and I'll send a follow-up.

> Dismissed @spaskob from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-:-M4J-OG-6cIfHNZ1AZfW:b-fn0mg7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/cmd/roachtest/autoupgrade.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ewu_Z3wBYalIHSfHm:-M4Iv3SA8W1l-FrFx77N:b-5gg6iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L47)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

For my understanding why do we test v2.1 - do we still have clients that we support running this version?
</blockquote></details>

Generally, as long as a release branch might see a new release cut from it, we will try to keep the roachtests runnable on them. Often, this is so tedious that we selectively skip some roachtests on old versions even though they haven't reached EOL yet. In this particular case, I think we do still have some kind of exception for v2.1, but the real reason you're still seeing this code is because this test hasn't been touched in a long time (I don't know if you realized, but the `autoupgrade.go` contents are pure movement - I haven't touched this test at all, though I will now in reaction to your comments).
___
*[pkg/cmd/roachtest/autoupgrade.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ex9tN9KxPkz4UIdiJ:-M4Iwe1CCbRqejq0IwKG:b-2mtfkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L70)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Do you mind running locally to check if this is still true.
</blockquote></details>

I don't think it is, or I would've needed it while reworking version-upgrade. Removed. (And tested)
___
*[pkg/cmd/roachtest/autoupgrade.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4Ey3-_4RZLSAUMu6LV:-M4IwiFg6DfOxzrwZPMP:be6bzdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L104)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

nit: Please add a comment on why we over-write the function arg here.
</blockquote></details>

I did add a comment to the best of my abilities (I didn't write or touch this test, or if I did then I forgot and rather not remember :-)) ).
___
*[pkg/cmd/roachtest/autoupgrade.go, line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4F0hy64epUQiWCg9Dd:-M4IwyXQ6DEDTzuNFpC7:b-pchieo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L134)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

what is the version of the new binary?
</blockquote></details>

it's the version we're testing, so if you're running on master it will be 20.2, if you're on release-19.1 is 19.1.x. Added a comment.
___
*[pkg/cmd/roachtest/autoupgrade.go, line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4F-6AA731Fl-fo2Vgs:-M4IxZoK5DmWR_v6qUvX:b-6yhd7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/autoupgrade.go#L170)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

nit: add a comment explaining what do we test by decomissioning this node.
</blockquote></details>

Adjusted the comment to make it clearer.
___
*[pkg/cmd/roachtest/versionupgrade.go, line 218 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47065#-M4F3BKhAhkPd5X9B1l5:-M4IvXiI3WiK3NcSAnOV:b-qky29o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f9cd27b7833406f57eb70230b0bb98714041430/pkg/cmd/roachtest/versionupgrade.go#L218)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Do you think these may be re-used by other mixed-version roachtests?
</blockquote></details>

Yes, the reason I am touching this code quite a bit is because I want to enable that, prompted by your use case.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=spaskob
